mantidproject/mantid
# Time validity for instrument parameters

**Description of work.**
It is now possible to set a validity range for each parameter in an instrument parameter file, by adding starting and ending times in the form `valid-from=AAAA-MM-DD HH:mm:SS valid-to=AAAA-MM-DD HH:mm:SS` (ISO format).

These flags are of course completely optional, and previously written IPF are still valid.

Moreover, some of ILL loaders had to be slightly modified, because they did not store the run time in an ISO compatible format. 

**To test:**

- In `ILL/D33_Parameters.xml`, add
```    
<parameter name="test-param" type="number">
      <value val="0" valid-to="2017-12-31T23:59:59" />
      <value val="1" valid-from="2018-01-01T00:00:00"/>
</parameter>
```
in the component-link d33.
- Open Workbench.
- Load `ExternalData/Data/UnitTest/ILL/D33/042610`.
- Run 
```
from mantid.simpleapi import *

ws = mtd["042610"]
inst = ws.getInstrument()
print("Does the instrument have the 'test-param' parameter ?")
print(inst.hasParameter("test-param"))
print("Is this parameter's value correct ?")
print(inst.getNumberParameter("test-param")[0] == 1.0)
```
and check the output.

- For a more comprehensive test, run the `LoadParameterFileTest` test.


Fixes #29505. 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
